### PR TITLE
Result classes for the http promise type

### PR DESCRIPTION
**Pull request by vpodzime**
~Requires https://github.com/cfengine/modules/pull/39 to be merged first. Only the last 3 commits are relevant for this PR.~

### CODE REVIEW

File: promise-types/http/README.org [hunk 1]
```diff
@@ -24,6 +24,23 @@ the URL is specified in the promiser.
...
+On success, each /http/ promise defines a class that is in the form of
+~$(this.promiser)_METHOD_request_done~. On failure, class of form
+~$(this.promiser)_METHOD_request_failed~ is defined plus classes of form:
+
+- ~$(this.promiser)_METHOD_payload_failed~ (payload-related failures)
+- ~$(this.promiser)_METHOD_payload_file_failed~ (file-based payload failures)
+- ~$(this.promiser)_METHOD_payload_conversion_failed~ (payload conversion failures)
+- ~$(this.promiser)_METHOD_file_failed~ (/file/-related failures, see above)
+
+in the respective cases. *The promiser is canonified in all the above classes.*
```
Comment (olehermanse) [inline promise-types/http/README.org:41]: What about failures related to the response code? Like a 400 client error, those are the ones I am most interested in.

File: promise-types/http/http_promise_type.py [hunk 2]
```diff
@@ -75,6 +75,8 @@ def evaluate_promise(self, promiser, attributes):
         target = attributes.get("file")
         insecure = attributes.get("insecure", False)
 
+        canonical_promiser = promiser.translate(str.maketrans({char: "_" for char in ("@", "/", ":", "?", "&", "%")}))
```
Comment (olehermanse) [inline promise-types/http/http_promise_type.py:78]: We could add this to cfengine.py

File: promise-types/http/README.org (hunk 1 above)
Comment (olehermanse) [reply promise-types/http/README.org:41]: I see that this will cause the generic one, but not a more specific class. Worth mentioning in the docs.

**Review by olehermanse — APPROVED** [review]
LGTM, just some smaller comments.

File: promise-types/http/http_promise_type.py (hunk 2 above)
Comment (vpodzime) [reply promise-types/http/http_promise_type.py:78]: Well, generally the promiser can have other various characters, right? And there's actually code on the cf-agent side that makes sure all the result classes are canonified. So this is really just for having nicer logs and such.

Comment (olehermanse) [reply promise-types/http/http_promise_type.py:78]: I meant we can provide a canonify function that you can use in cases like this.

**Pull request merged**